### PR TITLE
2.0

### PR DESCRIPTION
**Pull request by m2otech**
I have made some changes to the EMM software based on 1.2.1 sources. As the changes are not always incremental (e.g. some features have been removed, others added) I propose creating a new branch 2.0

The main changes are:
- removed scroll bar in each slot
- removed playlist feature (mostly commented out in code)
- added counter in each slot (how many times a slot was played)
- added Pause All / Stop All buttons (next to layer selection)
- added checkbox "Disable pausing this title" in slot properties
- modified keyboard key bindings and added new ones (e.g. for new feature Pause All and for opening Slot Store)
- small graphical changes (e.g. a paused title always has the original slot color)
- fixed a few small bugs (e.g. bottom-right slot was not cleared, when playing a slot the corresponding slot may have blinked in other layers too)

Most changes are marked with "m2" in the code.

Here a few screenshots of the changes:

emm-2 0 - new feature 4
emm-2 0 - new feature 1
emm-2 0 - new feature 2
emm-2 0 - new feature 3

### CODE REVIEW

**Comment by cevou** [comment]
Hi,

thanks for the pull request. I will review the change in the next couple of days and will add my feedback to the changed code.

Best regards,
Christoph

**Comment by m2otech** [comment]
Hi Christoph,

Thanks a lot. Let me know if I can help.

Best regards,
Marco

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Auto detect text files and perform LF normalization
+* text=auto
+
+# Custom for Visual Studio
```
Comment (cevou) [inline .gitattributes]: I don'T see a reason why we need all these extensions here. There are no .cs, .doc, et.c files in the project

File: main.cpp [hunk 2]
```diff
@@ -99,7 +99,16 @@ int main(int argc, char *argv[])
                 CartSlot *slot = AudioProcessor::getInstance()->getCartSlotWithNumber(number);
                 s->showMessage("Slots laden und überprüfen...\r\n"+slot->getText1());
                 if (slot->isMissing()) {
-                    QMessageBox::information(s,"Datei wurde nicht gefunden","Slot "+QString::number(slot->getNumber())+" ("+slot->getText1()+") konnte nicht geladen werden, weil die Datei nicht gefunden wurde!");
+                    //QMessageBox::information(s,"Datei wurde nicht gefunden","Slot "+QString::number(slot->getNumber())+" ("+slot->getText1()+") konnte nicht geladen werden, weil die Datei nicht gefunden wurde!");
```
Comment (cevou) [inline main.cpp]: If this is replaced by the new functionality, this old code can be removed

File: main.cpp [hunk 3]
```diff
@@ -99,7 +99,16 @@ int main(int argc, char *argv[])
...
+
+                    // m2: Ask user whether to ignore further "missing files"
+                    QMessageBox::StandardButton reply;
+                    reply = QMessageBox::question(s, QString::fromUtf8("Datei nicht gefunden"), "Slot "+QString::number(slot->getNumber())+" ("+slot->getText1()+") konnte nicht geladen werden, weil die Datei nicht gefunden wurde! \n \n Sollen weitere Meldungen angezeigt werden?", QMessageBox::Yes|QMessageBox::No);
+                    if (reply == QMessageBox::No) {
```
Comment (cevou) [inline main.cpp]: If youbreak all the loops here, all the following slots will not be preloaded anymore, which leads to worse performance. You should set a flag, that errors should be skipped and check against this flag in line 101

File: model/audio/abstractaudioobject.cpp [hunk 4]
```diff
@@ -139,18 +139,20 @@ void AbstractAudioObject::stop()
     unloadStream();
 }
 
-
 void AbstractAudioObject::pause()
 {
     if (playing)
     {
+        //qDebug("Pause called");
```
Comment (cevou) [inline model/audio/abstractaudioobject.cpp]: remove debug statements

File: model/audio/abstractaudioobject.cpp [hunk 5]
```diff
@@ -139,18 +139,20 @@ void AbstractAudioObject::stop()
...
         BASS_ChannelPause(stream);
         BASS_Mixer_ChannelFlags(stream, BASS_MIXER_PAUSE, BASS_MIXER_PAUSE);
         playing = false;
         paused = true;
+        emit pausedSignal();
     }
     else
     {
+        //qDebug("Pause (restart) called");
```
Comment (cevou) [inline model/audio/abstractaudioobject.cpp]: remove debug statements

File: model/audio/abstractaudioobject.h [hunk 6]
```diff
@@ -39,9 +39,11 @@ class AbstractAudioObject : public QObject
     bool isPlaying();
     bool isPaused();
     void fadeOut(int ms);
+    void fadeOutPause(int ms);
     bool isMissing();
     double getTimeLeft();
-    void pause();
+    // m2: redeclared as slot to call it asynchronously
```
Comment (cevou) [inline model/audio/abstractaudioobject.h]: remove old coding

File: model/audio/cartslot.cpp [hunk 7]
```diff
@@ -213,10 +239,21 @@ void CartSlot::readData()
...
         eqConfig = settings.value("Slot"+QString::number(number)+"/EqualizerConfig","").toString();
+        // m2: new checkbox disable pause (default false)
+        pauseDisabled = settings.value("Slot"+QString::number(number)+"/PauseDisabled",false).toBool();
     }
     else
     {
         QSqlQuery query;
+
+        // m2: Check if field pause_disabled exists, if not create it
+        query.prepare("SELECT pause_disabled FROM slots");
```
Comment (cevou) [inline model/audio/cartslot.cpp]: this should be check somewhere during startup of the slot store

File: model/clearlayerthread.cpp [hunk 8]
```diff
@@ -20,6 +20,8 @@
 #include "model/configuration.h"
 #include "view/cartslotwidget.h"
 
+#include "view/mainwindow.h"
```
Comment (cevou) [inline model/clearlayerthread.cpp]: why is mainwindow included here?

File: view/aboutdialog.ui [hunk 9]
```diff
@@ -57,7 +57,7 @@ font-weight:bold;</string>
 font-size:12px;</string>
         </property>
         <property name="text">
-         <string notr="true">1.2.1</string>
+         <string notr="true">2.1.0</string>
```
Comment (cevou) [inline view/aboutdialog.ui]: I don'T see this as version 2. More about this later.

**Comment by cevou** [comment]
Hi again,

I would prefer to have seperate features in separate pull request. That makes it easier to discuss the features and merge them seperately.
I don't see this as a 2.0 release. Generally functionality should not be removed. You can introduce configuration options to disable certain functionality.
For 2.0 I would recommend a complete restructuring of the coding. The application should be more modular and different modules should be available for slots, playlists, etc.

I would also prefer, that new functionality and changes are discussed here first, so we can agree on how things should be implemented or changed.

Best regards,
Christoph

**Comment by cevou** [comment]
You can also leave out all the comments, where you changed the coding. The changes can be seen in the diff.

File: model/clearlayerthread.cpp (hunk 8 above)
Comment (m2otech) [inline model/clearlayerthread.cpp]: I think I wanted to include some code that required that, then switched back. Can probably be removed.

File: main.cpp (hunk 3 above)
Comment (m2otech) [inline main.cpp]: Ok will look into that.

**Pull request closed without merging**